apps/wallet/src/engine/runners/solana/history.ts

```diff
@@ -0,0 +1,318 @@
+		metadata: await getMetadata(connection, mint),
+	};
+
+	return {
+		sender,
+		receiver,
+		amount: Math.abs(amount),
+		token,
+	};
+};
+
+const analyzeTokenBalances = (
+	preTokenBalances: TokenBalance[],
+	postTokenBalances: TokenBalance[],
+) => {
+	if (postTokenBalances.length > preTokenBalances.length) {
+		const referralPreBalance = preTokenBalances[0];
+		const correspondingPostBalance = postTokenBalances.find(
+			(balance) => balance.accountIndex === referralPreBalance.accountIndex,
+		);
+		const otherPostBalance = postTokenBalances.find(
+			(balance) => balance.accountIndex !== referralPreBalance.accountIndex,
+		);
+		const mint = referralPreBalance.mint;
+		const amount =
+			(correspondingPostBalance?.uiTokenAmount.uiAmount || 0) -
+			(referralPreBalance?.uiTokenAmount.uiAmount || 0);
+
+		let sender: string;
+		let receiver: string;
+		if (amount < 0) {
+			sender = correspondingPostBalance?.owner || '';
+			receiver = otherPostBalance?.owner || '';
+		} else {
+			sender = otherPostBalance?.owner || '';
+			receiver = correspondingPostBalance?.owner || '';
+		}
+
+		return {
+			mint,
+			amount,
+			sender,
+			receiver,
+		};
+	}
+
+	const referralPostBalance = postTokenBalances[0];
+	const correspondingPreBalance = preTokenBalances.find(
+		(balance) => balance.accountIndex === referralPostBalance.accountIndex,
+	);
+	const otherPreBalance = preTokenBalances.find(
+		(balance) => balance.accountIndex !== referralPostBalance.accountIndex,
+	);
+	const mint = referralPostBalance.mint;
+	const amount =
+		(referralPostBalance.uiTokenAmount.uiAmount || 0) -
+		(correspondingPreBalance?.uiTokenAmount.uiAmount || 0);
+
+	let sender: string;
+	let receiver: string;
+	if (amount < 0) {
+		sender = correspondingPreBalance?.owner || '';
+		receiver = otherPreBalance?.owner || '';
+	} else {
+		sender = otherPreBalance?.owner || '';
+		receiver = correspondingPreBalance?.owner || '';
+	}
+
+	return {
+		mint,
+		amount,
+		sender,
+		receiver,
+	};
+};
+
+const getTransactionFee = async (walletAddress: string) => {
+	return await storage.safeGet<TokenDocument>(
+		`${walletAddress}/token/${solana.solMint}`,
+	);
+};
+
+const removeOldHistories = async (
+	newHistoryDocs: TransactionHistoryDocument[],
+) => {
+	const { docs } = await storage.find(selectors.allHistories);
+
+	if (docs.length === 0) return;
+
+	const oldHistoryDocs = (docs as TransactionHistoryDocument[]).sort(
+		(doc1, doc2) => (doc1.date > doc2.date ? -1 : 1),
+	);
+
+	for (let i = 0; i < 20; i++) {
+		if (newHistoryDocs.length === 0 && oldHistoryDocs.length === 0) return;
+
+		const currentNewDoc = newHistoryDocs[0];
+		const currentOldDoc = oldHistoryDocs[0];
+
+		if (currentNewDoc) {
+			if (currentNewDoc._id === currentOldDoc._id) {
+				oldHistoryDocs.shift();
+			}
+			newHistoryDocs.shift();
+		} else {
+			oldHistoryDocs.shift();
+		}
+	}
+
+	Promise.all(
+		oldHistoryDocs.map(async (doc) => {
+			await storage.removeDoc(doc._id);
+		}),
+	);
```

Comment:
if using `Promise.all`, It should be aware that it needs to use await to have correct behavior, If a function calls and waits for this one, it will not wait for complete removing